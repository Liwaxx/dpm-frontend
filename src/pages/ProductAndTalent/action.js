import { ACTIONS } from '../../constants';
import { SERVICES, HEADERS } from '../../configs';
import fetch from '../../utils/fetch';

export function fetchProduct() {
  return (dispatch) => {
    const options = {
      method: 'get',
      url: `${SERVICES.GET_PRODUCT}`,
      headers: HEADERS.HEADERS(),
    };

    dispatch(loadingProductAction());

    fetch(options)
      .then(response => {
        dispatch(dataProductFetched(response.data));
      })
      .catch(() => {
        dispatch(dataProductFetched([]));
      });
  };
}

export function fetchQueueAllProject(payload) {
  return (dispatch) => {
    const options = {
      method: 'get',
      url: `${SERVICES.GET_QUEUE_ALL_PROJECT}/${payload}`,
      headers: HEADERS.HEADERS(),
    };

    dispatch(loadingAction());
    
    fetch(options)
      .then(response => {
        dispatch(dataAllProjectFetched(response.data));
      })
      .catch(() => {
        dispatch(dataAllProjectFetched([]));
      });
  };
}

export function fetchDataSquadPerformance(payload = {}) {
  return (dispatch) => {
    const options = {
      method: 'get',
      url: `${SERVICES.GET_SQUAD_PERFORMANCE}/${payload.user}/${payload.year}`,
      headers: HEADERS.HEADERS(),
    };

    dispatch(loadingAction());
    
    fetch(options)
      .then(response => {
        dispatch(dataSquadPerformanceFetched(response.data));
      })
      .catch(() => {
        dispatch(dataSquadPerformanceFetched([]));
      });
  };
}

export function fetchDataTalentPerformance() {
  return (dispatch) => {
    const options = {
      method: 'get',
      url: `${SERVICES.GET_TALENT_PERFORMANCE}`,
      headers: HEADERS.HEADERS(),
    };

    dispatch(loadingProductAction());

    fetch(options)
      .then(response => {
        dispatch(dataTalentPerformanceFetched(response.data));
      })
      .catch(() => {
        dispatch(dataTalentPerformanceFetched([]));
      });
  };
}

function loadingAction() {
  return { type: ACTIONS.LOADING };
}

function loadingProductAction() {
  return { type: ACTIONS.LOADING_PRODUCT };
}

function dataProductFetched(dataProduct) {
  return {
    type: ACTIONS.LIST_PRODUCT_FETCHED,
    dataProduct,
  };
}

function dataAllProjectFetched(data) {
  return {
    type: ACTIONS.LIST_ALL_PROJECT,
    dataAllProject: data,
  };
}

function dataSquadPerformanceFetched(data) {
  return {
    type: ACTIONS.LIST_OF_SQUAD_PERFORMANCE,
    dataSquadPerformance: data,
  };
}

function dataTalentPerformanceFetched(data) {
  return {
    type: ACTIONS.LIST_OF_TALENT_PERFORMANCE,
    dataTalentPerformance: data,
  };
}