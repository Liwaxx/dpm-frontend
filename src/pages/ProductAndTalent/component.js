import React from 'react';
import CircularProgress from '@material-ui/core/CircularProgress';
import PropTypes from 'prop-types';
import Card from '../../components/elements/Card';
import Linechart from '../../components/elements/LineChart';
import TalentQueueTable from '../../components/elements/TalentQueueTable';
import TalentScoreCard from '../../components/elements/TalentScoreCard';
import TalentVacancyCard from '../../components/elements/TalentVacancyCard';
import Text from '../../components/elements/Text';

const scores = [
  {
    name: 'Backend Developer',
    rank: [{ name: 'Alif Septian Nurdianto', previous: 2, point: 5 }, { name: 'Irman Juliansyah', previous: 5, point: 4.8 }, { name: 'Kevin Warsahid', previous: 1, point: 4.6 }]
  },
  {
    name: 'Frontend Web Developer',
    rank: [{ name: 'Emmy Trinawati H', previous: 3, point: 5 }, { name: 'Dama Samudera', previous: 2, point: 4.9 }, { name: 'Rozinul Mahrus', previous: 4, point: 4.7 }]
  },
  {
    name: 'Frontend Mobile Developer',
    rank: [{ name: 'Rizky Ramadhan S', previous: 1, point: 5 }, { name: 'Egi Ari Wibowo', previous: 4, point: 4.7 }, { name: 'Eko Meidianto', previous: 2, point: 4.6 }]
  },
  {
    name: 'QA Engineer',
    rank: [{ name: 'Aldila Karina', previous: 2, point: 5 }, { name: 'Dini Indriyani P', previous: 1, point: 4.7 }, { name: 'Andria Sufy', previous: 5, point: 4.5 }]
  },
  {
    name: 'Document Engineer',
    rank: [{ name: 'Elnoor Fajrah', previous: 4, point: 5 }, { name: 'Riska Desrianti', previous: 2, point: 4.9 }, { name: 'Edwina Anky', previous: 1, point: 4.5 }]
  },
  {
    name: 'Code Reviewer',
    rank: [{ name: 'Oka Uliandana', previous: 2, point: 5 }, { name: 'Alif Septian Nurdianto', previous: 1, point: 4.9 }, { name: 'Muhammad Ilham R', previous: 3, point: 4.9 }]
  },
];

const vacancies = [
  { name: 'UX Designer Lead', floating: 70, vacant: 30 },
  { name: 'UX Designer Senior', floating: 80, vacant: 20 },
  { name: 'QA Junior', floating: 30, vacant: 70 },
  { name: 'Front End Medium', floating: 10, vacant: 90 },
  { name: 'Partnership Lead', floating: 0, vacant: 100 },
];

const dataOnGoing = [
  { name: 'Logistik Tani', version: '1.0' },
  { name: 'QRen 2.0', version: '2.4' },
  { name: 'Bulir', version: '2.2' },
  { name: 'Dialer SLI 01017', version: '1.3' },
];

const grafikQueue = [
  { month: 'October', onGoingProject: 0, queueProject: 0 },
  { month: 'December', onGoingProject: 0, queueProject: 0 },
  { month: 'Januari', onGoingProject: 0, queueProject: 0 }
];

export default class Component extends React.Component {
  componentDidMount() {
    const { actions } = this.props;

    actions.fetchProduct();
    actions.fetchQueueAllProject('2018');
    actions.fetchDataSquadPerformance({ user: 'P001', year: '2018' });
  }

  _renderSectionHeader(title) {
    const { classes } = this.props;

    return (
      <div className={classes.sectionHeader}>
        <Text type="display">{title}</Text>
        <div className={classes.sectionActions} />
      </div>
    );
  }

  _renderSectionOngoing() {
    const { classes, isLoadingProduct } = this.props;

    return (
      <Card className={classes.ongoing}>
        {this._renderSectionHeader('On Going')}
        <div className={classes.vacanciesContent}>
          {isLoadingProduct ? <CircularProgress /> : <TalentQueueTable data={dataOnGoing} />}
        </div>
      </Card>
    );
  }

  _renderSectionQueue() {
    const { classes, isLoadingProduct } = this.props;

    return (
      <Card className={classes.queue}>
        {this._renderSectionHeader('Queue')}
        <div className={classes.vacanciesContent}>
          {isLoadingProduct ? <CircularProgress /> : <TalentQueueTable data={[]} />}
        </div>
      </Card>
    );
  }

  _renderSectionAllQueue() {
    const { classes = [], } = this.props;
    const legend = ['Download', 'User Aktif'];
    let data = [['month', 'onGoingProject', 'queueProject']];

    let dataAllProject = grafikQueue;

    if( dataAllProject && dataAllProject.length >= 0 ) {
      dataAllProject.map(function(item) {
        data.push([item.month, item.onGoingProject, item.queueProject]);
      });
    }

    return (
      <Card className={classes.allQueue}>
        {this._renderSectionHeader('Queue All Project')}
        <div className={classes.vacanciesContent}>
          { dataAllProject.length >= 1 ? 
            <Linechart data={data} legend={legend} /> : 
            <CircularProgress />}
        </div>
      </Card>
    );
  }

  _renderSectionScores() {
    const { classes } = this.props;

    return (
      <Card className={classes.scores}>
        {this._renderSectionHeader('Talent Performance  Oktober 2018')}
        <div className={classes.scoresContent}>
          {scores.map((item, idx) => (
            <TalentScoreCard className={classes.score} data={item} key={idx} />
          ))}
        </div>
      </Card>
    );
  }

  _renderSectionSquad() {
    const { classes, dataSquadPerformance = [], } = this.props;
    let legend = ['Squad Bulir'];
    let data = [['name', 'point']];

    if( dataSquadPerformance && dataSquadPerformance.length >= 1 ) {
      dataSquadPerformance.map(function(item) {
        data.push([item.name, item.point]);
      });
    }

    return (
      <Card className={classes.squad}>
        {this._renderSectionHeader('Squad Performance')}
        <div className={classes.vacanciesContent}>
          {dataSquadPerformance.length >= 1 ? 
            <Linechart data={data} legend={legend} /> : 
            <CircularProgress />}
        </div>
      </Card>
    );
  }

  _renderSectionVacancies() {
    const { classes } = this.props;

    return (
      <Card className={classes.vacancies}>
        {this._renderSectionHeader('Talent')}
        <div className={classes.vacanciesContent}>
          {vacancies.map((item, idx) => (
            <TalentVacancyCard className={classes.vacancy} data={item} key={idx} />
          ))}
        </div>
      </Card>
    );
  }

  render() {
    const { classes } = this.props;

    return (
      <div className={classes.root}>
        <div className={classes.rowTop}>
          {this._renderSectionQueue()}
          {this._renderSectionOngoing()}
          {this._renderSectionAllQueue()}
          {this._renderSectionSquad()}
        </div>
        <div className={classes.rowBottom}>
          {this._renderSectionScores()}
          {this._renderSectionVacancies()}
        </div>
      </div>
    );
  }
}

Component.propTypes = {
  actions: PropTypes.object,
  classes: PropTypes.object,
  dataAllProject: PropTypes.array,
  dataSquadPerformance: PropTypes.array,
  dataProduct: PropTypes.array,
  isLoadingProduct: PropTypes.bool,
  isLoading: PropTypes.bool,
};