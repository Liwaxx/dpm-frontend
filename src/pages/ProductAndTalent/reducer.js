import { ACTIONS } from '../../constants';

const initialState = {
  isLoadingProduct: false,
  isLoading: false,
  dataProduct: [],
  dataAllProject: [],
  dataSquadPerformance: [],
};

export default function reducer(state = initialState, action) {
  const { type, dataProduct, dataAllProject, dataSquadPerformance } = action;
  switch (type) {
    case ACTIONS.LOADING:
      return {
        ...state,
        isLoading: true
      };
    case ACTIONS.LOADING_PRODUCT:
      return {
        ...state,
        isLoadingProduct: true
      };
    case ACTIONS.LIST_PRODUCT_FETCHED:
      return {
        ...state,
        isLoadingProduct: false,
        dataProduct,
      };
    case ACTIONS.LIST_ALL_PROJECT:
      return {
        ...state,
        isLoading: false,
        dataAllProject,
      };
    case ACTIONS.LIST_OF_SQUAD_PERFORMANCE:
      return {
        ...state,
        isLoading: false,
        dataSquadPerformance,
      };
    default:
      return state;
  }
}