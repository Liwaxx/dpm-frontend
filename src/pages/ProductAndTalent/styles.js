const styles = {
  allQueue: {
    flexGrow: 1,
    height: 'auto',
    marginRight: '16px',
  },
  ongoing: {
    flexGrow: 1,
    height: 'auto',
    marginRight: '16px',
    overflow: 'auto',
  },
  queue: {
    flexGrow: 1,
    height: 'auto',
    marginRight: '16px',
    overflow: 'auto',
  },
  root: {
    display: 'flex',
    flexDirection: 'column',
    height: '100vh',
  },
  rowBottom: {
    display: 'flex',
    flexBasis: '70%',
  },
  rowTop: {
    display: 'flex',
    flexBasis: '30%',
    marginBottom: '16px',
  },
  score: {
    flexBasis: '32%',
    marginBottom: '16px',
  },
  scores: {
    flexGrow: 5,
    height: 'auto',
    marginRight: '16px',
    overflow: 'auto',
    width: 'min-content',
  },
  scoresContent: {
    display: 'flex',
    flexWrap: 'wrap',
    justifyContent: 'space-between',
    padding: '16px 20px',
  },
  sectionActions: {
    display: 'flex',
    justifyContent: 'flex-end',
  },
  sectionHeader: {
    backgroundColor: '#ffffff',
    borderLeft: '4px solid #54baed',
    display: 'flex',
    justifyContent: 'space-between',
    padding: '16px 20px 0',
    position: 'sticky',
    top: 0,
  },
  squad: {
    flexGrow: 1,
    height: 'auto',
  },
  vacancies: {
    flexGrow: 1,
    height: 'auto',
    overflow: 'auto',
  },
  vacanciesContent: {
    padding: '16px 20px',
  },
  vacancy: {
    marginBottom: '16px',
  },
};

export default styles;