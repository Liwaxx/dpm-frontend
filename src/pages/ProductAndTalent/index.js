import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { withStyles } from '@material-ui/core/styles';
import Component from './component';
import * as actions from './action';
import styles from './styles';

function mapStateToProps(state) {
  const { 
    isLoading, 
    isLoadingProduct, 
    dataProduct, 
    dataAllProject, 
    dataSquadPerformance, 
  } = state.talent;
  
  return {
    isLoading,
    dataProduct,
    isLoadingProduct,
    dataAllProject,
    dataSquadPerformance,
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators(actions, dispatch)
  };
}

const StyledHome = withStyles(styles)(Component);

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(StyledHome);