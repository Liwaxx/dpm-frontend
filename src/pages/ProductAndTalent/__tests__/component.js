import React from 'react';
import renderer from 'react-test-renderer';
import ProductAndTalent from '../component';

jest.mock('../../../components/elements/Card');
jest.mock('../../../components/elements/TalentQueueTable');
jest.mock('../../../components/elements/TalentScoreCard');
jest.mock('../../../components/elements/TalentVacancyCard');
jest.mock('../../../components/elements/Text');

describe('ProductAndTalent', () => {
  let props = {};
  beforeEach(() => {
    props = {
      actions: { 
        fetchProduct: jest.fn(),
        fetchQueueAllProject: jest.fn(),
        fetchDataSquadPerformance: jest.fn(),
      },
      classes: {},
      isLoadingProduct: false,
    };
  });

  it('renders correctly', () => {
    const tree = renderer
      .create(<ProductAndTalent {...props} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders loading state', () => {
    props.isLoadingProduct = true;
    const tree = renderer
      .create(<ProductAndTalent {...props} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});