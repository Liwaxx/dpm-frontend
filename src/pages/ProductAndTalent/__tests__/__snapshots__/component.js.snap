// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductAndTalent renders correctly 1`] = `
<div>
  <div />
  <div />
</div>
`;

exports[`ProductAndTalent renders loading state 1`] = `
<div>
  <div />
  <div />
</div>
`;
